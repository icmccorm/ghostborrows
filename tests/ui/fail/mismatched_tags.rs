use ghostborrows::*;

fn main() {
    let value = Value::new(1);
    value.borrow(|r1| {
        r1.borrow(|r2| {
            let (ptr1, _) = r1.split();
            let (_, perm2) = r2.split();
            let _ = Ref {
                pointer: ptr1,
                permission: perm2,
            };
        })
    });
}
