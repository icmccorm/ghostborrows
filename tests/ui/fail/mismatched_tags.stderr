error[E0521]: borrowed data escapes outside of closure
 --> tests/ui/fail/mismatched_tags.rs:8:30
  |
5 |     value.borrow(|r1| {
  |                   -- `r1` declared here, outside of the closure body
6 |         r1.borrow(|r2| {
  |                    -- `r2` is a reference that is only valid in the closure body
7 |             let (ptr1, _) = r1.split();
8 |             let (_, perm2) = r2.split();
  |                              ^^^^^^^^^^ `r2` escapes the closure body here
  |
  = note: requirement occurs because of the type `ghostborrows::Ref<'_, i32>`, which makes the generic argument `'_` invariant
  = note: the struct `ghostborrows::Ref<'tag, T>` is invariant over the parameter `'tag`
  = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error[E0521]: borrowed data escapes outside of closure
 --> tests/ui/fail/mismatched_tags.rs:7:29
  |
5 |     value.borrow(|r1| {
  |                   --
  |                   |
  |                   `r1` is a reference that is only valid in the closure body
  |                   has type `ghostborrows::Ref<'1, i32>`
6 |         r1.borrow(|r2| {
7 |             let (ptr1, _) = r1.split();
  |                             ^^^^^^^^^^
  |                             |
  |                             `r1` escapes the closure body here
  |                             argument requires that `'1` must outlive `'static`
